pkg/limits/frontend/ring_test.go

```diff
@@ -10,227 +10,379 @@ import (
+			Streams: []*logproto.StreamMetadata{{
+				StreamHash:             0x2,
+				EntriesSize:            0x4,
+				StructuredMetadataSize: 0x5,
+			}},
+		}, {
+			Tenant: "test",
+			Streams: []*logproto.StreamMetadata{{
+				StreamHash:             0x1,
+				EntriesSize:            0x2,
+				StructuredMetadataSize: 0x3,
+			}},
+		}},
+		exceedsLimitsResponses: []*logproto.ExceedsLimitsResponse{{
+			Tenant: "test",
+			Results: []*logproto.ExceedsLimitsResult{{
+				StreamHash: 0x2,
+				Reason:     uint32(limits.ReasonExceedsMaxStreams),
+			}},
+		}, {
+			Tenant: "test",
+			Results: []*logproto.ExceedsLimitsResult{{
+				StreamHash: 0x1,
+				Reason:     uint32(limits.ReasonExceedsMaxStreams),
+			}},
+		}},
+		exceedsLimitsResponseErrs: []error{nil, nil},
+		expected: []*logproto.ExceedsLimitsResponse{{
+			Tenant: "test",
+			Results: []*logproto.ExceedsLimitsResult{{
+				StreamHash: 0x1,
+				Reason:     uint32(limits.ReasonExceedsMaxStreams),
+			}},
+		}, {
+			Tenant: "test",
+			Results: []*logproto.ExceedsLimitsResult{{
+				StreamHash: 0x2,
+				Reason:     uint32(limits.ReasonExceedsMaxStreams),
+			}},
+		}},
+	}, {
+		// When one instance returns an error, the entire request is failed.
+		name: "two streams, two instances, one instance returns error",
+		request: &logproto.ExceedsLimitsRequest{
+			Tenant: "test",
+			Streams: []*logproto.StreamMetadata{{
+				StreamHash:             0x1, // 0x1 is assigned to partition 1.
+				EntriesSize:            0x2,
+				StructuredMetadataSize: 0x3,
+			}, {
+				StreamHash:             0x2, // 0x2 is also assigned to partition 0.
+				EntriesSize:            0x4,
+				StructuredMetadataSize: 0x5,
+			}},
+		},
+		instances: []ring.InstanceDesc{{
+			Addr: "instance-0",
+		}, {
+			Addr: "instance-1",
+		}},
+		numPartitions:                     2,
+		expectedAssignedPartitionsRequest: []*logproto.GetAssignedPartitionsRequest{{}, {}},
+		getAssignedPartitionsResponses: []*logproto.GetAssignedPartitionsResponse{{
+			AssignedPartitions: map[int32]int64{
+				0: time.Now().UnixNano(),
+			},
+		}, {
+			AssignedPartitions: map[int32]int64{
+				1: time.Now().UnixNano(),
+			},
+		}},
+		expectedExceedsLimitsRequests: []*logproto.ExceedsLimitsRequest{{
+			Tenant: "test",
+			Streams: []*logproto.StreamMetadata{{
+				StreamHash:             0x2,
+				EntriesSize:            0x4,
+				StructuredMetadataSize: 0x5,
+			}},
+		}, {
+			Tenant: "test",
+			Streams: []*logproto.StreamMetadata{{
+				StreamHash:             0x1,
+				EntriesSize:            0x2,
+				StructuredMetadataSize: 0x3,
+			}},
+		}},
+		exceedsLimitsResponses: []*logproto.ExceedsLimitsResponse{{
+			Tenant: "test",
+			Results: []*logproto.ExceedsLimitsResult{{
+				StreamHash: 0x2,
+				Reason:     uint32(limits.ReasonExceedsMaxStreams),
+			}},
+		}, nil},
+		exceedsLimitsResponseErrs: []error{nil, errors.New("an unexpected error occurred")},
+		expectedErr:               "an unexpected error occurred",
```

Comment:
This will change once we figure out what to do when we cannot get an answer for a partition.